controllers/imagerepository_controller.go

```diff
@@ -230,6 +234,139 @@ func (r *ImageRepositoryReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+					// Updated item in Spec.Notifications: update notification in Quay
+					log.Info("Updating notification in Quay", "Title", notification.Title, "Event", notification.Event, "Method", notification.Method)
+					updatedNotification, err := r.QuayClient.UpdateNotification(
+						r.QuayOrganization,
+						imageRepository.Spec.Image.Name,
+						statusNotification.UUID,
+						quay.Notification{
+							Title:  notification.Title,
+							Event:  string(notification.Event),
+							Method: string(notification.Method),
+							Config: quay.NotificationConfig{
+								Url: notification.Config.Url,
+							},
+							EventConfig: quay.NotificationEventConfig{},
+						})
+					if err != nil {
+						log.Error(err, "failed to update notification", "Title", statusNotification.Title, "UUID", statusNotification.UUID)
+						return err
+					}
+					statusNotification.UUID = updatedNotification.UUID
+				}
+			} else {
+				quayNotification, err := r.NotificationExistsInQuayByTitle(notification.Title, imageRepository)
+				if err != nil {
+					return err
+				}
+				if quayNotification.UUID == "" {
+					// New item in Spec.Notifications: add notification to Quay and Status.Notifications
+					log.Info("Adding new notification to Quay", "Title", notification.Title, "Event", notification.Event, "Method", notification.Method)
+					notificationStatus, err := r.AddNotification(notification, imageRepository)
+					if err != nil {
+						log.Error(err, "failed to add notification", "Title", notification.Title, "Event", notification.Event, "Method", notification.Method)
+						return err
+					}
+					imageRepository.Status.Notifications = append(imageRepository.Status.Notifications, notificationStatus)
+				}
+			}
+		}
+		if existsInSpec == false {
+			// Deleted item from Spec.Notifications: delete notification from Quay and Status.Notifications
+			log.Info("Deleting notification in Quay", "Title", statusNotification.Title, "UUID", statusNotification.UUID)
+			_, err := r.QuayClient.DeleteNotification(
+				r.QuayOrganization,
+				imageRepository.Spec.Image.Name,
+				statusNotification.UUID)
+			if err != nil {
+				log.Error(err, "failed to delete notification", "Title", statusNotification.Title, "UUID", statusNotification.UUID)
+				return err
+			}
+			// Remove notification from CR status
+			imageRepository.Status.Notifications = append(imageRepository.Status.Notifications[:index], imageRepository.Status.Notifications[index+1:]...)
+		}
+	}
+
+	if err := r.Client.Status().Update(ctx, imageRepository); err != nil {
+		log.Error(err, "failed to update image repository status")
+		return err
+	}
+
+	return nil
+}
+
+func (r *ImageRepositoryReconciler) AddNotification(notification imagerepositoryv1alpha1.Notifications, imageRepository *imagerepositoryv1alpha1.ImageRepository) (imagerepositoryv1alpha1.NotificationStatus, error) {
```

Comment:
Should we do any validation before sending a request to Quay?

Reply:
Let me check that with Ales. Here, 100% I just refactored the code that was already written.